### PR TITLE
fix(spdx): guard against panic on empty license expressions

### LINKED ISSUE #2968: osv-scanner panics on empty license field (index out of range) when --licenses is enabled
### Summary

`osv-scanner scan --licenses <allowlist>` crashes with an unrecoverable panic (`index out of range`) when scanning packages whose license field is empty. An empty license string reaches `spdx.Satisfies("", allowlist)`, the tokenizer produces an empty token list, and `tokens.next()` dereferences `ts.tokens[0]` without a bounds check.

This is reachable from **valid parser output**, not malformed input: the APK extractor emits `Licenses: []string{record["L"]}` and, when an Alpine package record omits the `L:` field, that yields `[]string{""}`. The license enricher preserves it and it flows into the SPDX evaluator. With no `recover()` anywhere in the scan path, the panic aborts the whole scan.

### Affected code

`internal/spdx/satisfies.go:72-77`:
```go
// next returns the next token in the list of tokens, removing it from the list in the process
func (ts *tokens) next() string {
	token := ts.tokens[0]
	ts.tokens = ts.tokens[1:]

	return token
}
```
No bounds check exists before `ts.tokens[0]`, unlike `peek()` (`:63-69`) which does guard with `len(ts.tokens) == 0`.

Call chain: `Satisfies("", ...)` -> `parse` (`:157`) -> `parseOr` -> `parseAnd` -> `parseExpression` (`:216`) -> `nextAndIsNextNextValid` (`:92`) -> `next` (`:73`) -> panic.

### Reproduction

At HEAD `d87a911` (go 1.26+):

```bash
go test ./internal/spdx -run TestPocSatisfiesEmptyLicense -v
```

Observed output (`poc_satisfies_test.go` calling `spdx.Satisfies("", []string{"MIT"})`):
```
panic: runtime error: index out of range [0] with length 0
github.com/google/osv-scanner/v2/internal/spdx.(*tokens).next(...)
        C:/.../internal/spdx/satisfies.go:73
```

End-to-end: scan an Alpine repository/container image containing a package whose `lib/apk/db/installed` record omits the `L:` field with `osv-scanner scan --licenses MIT <target>`. The empty license reaches `spdx.Satisfies("")` and the process terminates.

### Impact

- Complete denial of the scan whenever `--licenses` is enabled and any scanned package lacks a license field.
- CI pipelines fail consistently until the package or flag is removed.
- Remaining packages in the scan are never evaluated.

### Root cause / suggested fix

- `tokenise` can legitimately produce zero tokens (empty input). `next()` must not index `ts.tokens[0]` unconditionally - return an error (or EOF token) when the stream is exhausted.
- Skip empty license strings before evaluating (e.g. treat `""` as no license / not satisfied rather than crashing).
- Add regression tests for empty license expressions.
- Consider a top-level `recover()` around scan execution so one malformed package record cannot terminate the whole process.

### PR DESCRIPTION
**Pull request by sl4x0**
## Overview

Fixes #2968

`osv-scanner scan --licenses <allowlist>` panics with `index out of range` when any scanned package has an empty license field. An empty license string reaches `spdx.Satisfies("", allowlist)`: the tokenizer produces zero tokens, and `tokens.next()` dereferences `ts.tokens[0]` without a bounds check, terminating the whole scan.

## Details

Two changes in `internal/spdx/satisfies.go`:

1. `nextAndIsNextNextValid` now returns an error when the token stream is exhausted instead of calling `next()` on an empty list, so an empty expression can never reach the index-out-of-range dereference. This is the root-cause guard (`peek()` already protected itself; `next()` did not).
2. `Satisfies` handles empty/whitespace-only expressions explicitly, returning `false, nil` so an empty license is treated as "not satisfied by the allowlist" and recorded as a violation (via `pkg/osvscanner/vulnerability_result.go`) rather than aborting the scan.

Regression tests:

- `internal/spdx/satisfies_test.go`: empty (`""`) and whitespace-only (`"   "`) license expressions added to `TestSatisfies`, asserting they evaluate to not-satisfied without an error (these panicked at HEAD).
- `pkg/osvscanner/vulnerability_result_internal_test.go`: `TestBuildVulnerabilityResultsWithEmptyLicense` drives a package with `Licenses: []string{""}` through `buildVulnerabilityResults` with a `MIT` allowlist and asserts the empty license is reported as a violation and the scan completes.

## Testing

- Reproduced the panic at HEAD with `go test ./internal/spdx -run TestSatisfies` before the fix.
- `go test ./internal/spdx/... ./pkg/osvscanner/...` passes after the fix (including the new regression tests).
- `go vet ./internal/spdx/... ./pkg/osvscanner/...` reports no issues.
- `go build ./...` succeeds.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/).
- [x] I have read the [CONTRIBUTING guide](https://github.com/google/osv-scanner/blob/main/CONTRIBUTING.md), and understand when to open a pull request
- [ ] I have run the linter using `./scripts/run_lints.sh`.
- [x] I have run the unit tests using `./scripts/run_tests.sh`.
- [x] I have made my commits and PR title follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.

### CODE REVIEW

**Review by G-Rath — CHANGES_REQUESTED** [review]
can you tone the comments down a bit? we don't need to be describing the past behaviour because that's what git and commit messages are for

**Comment by sl4x0** [comment]
Note on the checklist: the linter (\./scripts/run_lints.sh\ / golangci-lint v2.11.4) flags 19 files in the \osvscanner\ package as gofmt-unformatted on my Windows checkout, including files this PR does not touch (e.g. \pkg/osvscanner/stats.go\, \pkg/osvscanner/vulnerability_result.go\). Root cause is git's CRLF autocrlf conversion on Windows, not the diff. I verified the changed files (and an untouched control file) are gofmt-clean by running gofmt on LF-normalized copies, and CI on Linux will confirm.

File: internal/spdx/satisfies_test.go [hunk 1]
```diff
@@ -79,6 +79,23 @@ func TestSatisfies(t *testing.T) {
 				{"Apache-2.0"},
 			},
 		},
+		// empty expression licenses contain no tokens and can never satisfy an
+		// allowlist, but must not panic (regression for #2968)
```
Comment (another-rex) [inline internal/spdx/satisfies_test.go]:
```suggestion
		// empty expression licenses contain no tokens and can never satisfy an
		// allowlist.
```

File: pkg/osvscanner/vulnerability_result_internal_test.go [hunk 2]
```diff
@@ -175,3 +177,46 @@ func Test_assembleResult(t *testing.T) {
 		})
 	}
 }
+
+// TestBuildVulnerabilityResultsWithEmptyLicense ensures a package with an empty
+// license field does not panic when a license allowlist is configured, and is
+// instead recorded as a license violation (regression for #2968).
```
Comment (another-rex) [inline pkg/osvscanner/vulnerability_result_internal_test.go]:
```suggestion
// TestBuildVulnerabilityResultsWithEmptyLicense ensures a package with an empty
// license field does not panic when a license allowlist is configured, and is
// instead recorded as a license violation.
```

**Review by another-rex — APPROVED** [review]
